### PR TITLE
make viz preview live

### PR DESCRIPTION
**Pull request by stushurik**

### CODE REVIEW

File: packages/neoFrontend/src/VizPreview/LiveVizPreview.jsx [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import React, { useCallback } from 'react';
+import { useValue } from '../useValue';
+import { VizPreview } from '.';
+
+export const LiveVizPreview = ({ vizInfo$, getUser }) => {
+  const getOwner = useCallback(info => info.owner, []);
```
Comment (curran) [inline packages/neoFrontend/src/VizPreview/LiveVizPreview.jsx]: This can be factored out. No need for `useCallback` here.

Comment (curran) [reply packages/neoFrontend/src/VizPreview/LiveVizPreview.jsx]: Please try `const getOwner = info => info.owner;`, outside this closure.

**Review by curran — CHANGES_REQUESTED** [review]
Still reviewing, but wanted to give you this first feedback.

File: packages/neoFrontend/src/VizPreview/LiveVizPreview.jsx (hunk 1 above)
Comment (curran) [reply packages/neoFrontend/src/VizPreview/LiveVizPreview.jsx]: Better yet, move to where other accessors are defined.

File: packages/neoFrontend/src/VizPreview/VizPreview.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+          backgroundImage: `url(/api/visualization/thumbnail/${id}.png)`,
+        }}
+      />
+      <VizPreviewFooter borderRadiusLarge={true}>
+        <VizPreviewTitle to={link}>{title}</VizPreviewTitle>
+        <Bottom>
+          <Author ownerUser={ownerUser} isSmall={true} />
+          <Voter
+            canVote={false}
+            didVote={false}
```
Comment (curran) [inline packages/neoFrontend/src/VizPreview/VizPreview.js:42]: Nice. This part here is ripe for extension to support the actual upvoting. Just need to pull in the right contexts. Hooray!

File: packages/neoFrontend/src/useSameRef.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { useEffect, useRef } from 'react';
+
+export const useSameRef = (value) => {
```
Comment (curran) [inline packages/neoFrontend/src/useSameRef.js:3]: An interesting utility.

File: packages/neoFrontend/src/vizRealTimeHooks/useMultiOpStreams.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import { useMemo, useEffect } from 'react';
+import { Subject } from 'rxjs';
+import { useSameRef } from '../useSameRef';
+
+// Subscribes to a ShareDB document for ops,
+// returns an RxJS Subject that emits values for each op.
+export const useMultiOpStreams = (shareDBDocsToTrack = [], getPrevious) => {
+  const shareDBDocs = useSameRef(shareDBDocsToTrack);
```
Comment (curran) [inline packages/neoFrontend/src/vizRealTimeHooks/useMultiOpStreams.js:8]: Very nice how `shareDBDocsToTrack` is decoupled here. Paves the way to potentially drive these with client-side ShareDB queries.

File: packages/neoFrontend/src/vizRealTimeHooks/useShareDBQuery.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  useEffect(() => {
+    let query = null;
+
+    if (!connection || ids.length === 0) return;
+
+    // Clear out old doc in case ids changed.
+    // (don't want the user to see stale stuff).
+    setShareDBDocs(RESET_DOCS);
+    
+    query = connection.createSubscribeQuery(collection, {id: { $in: ids}}, null, (err, docs) => {
```
Comment (curran) [inline packages/neoFrontend/src/vizRealTimeHooks/useShareDBQuery.js:33]: Amazing!!! It does use client-side ShareDB query! Great job here.

File: packages/neoFrontend/src/VizPreview/LiveVizPreview.jsx (hunk 1 above)
Comment (stushurik) [reply packages/neoFrontend/src/VizPreview/LiveVizPreview.jsx]: @curran done

**Review by curran — CHANGES_REQUESTED** [review]
@stushurik Amazing work here!!!

However, it's broken under the condition of infinite scroll. Deploying this would introduce a regression of the infinite scroll feature. Please focus efforts to preserving infinite scroll and add more commits to this PR. Thanks!

**Comment by curran** [comment]
Wait... I may be wrong... sorry. Hang on.

**Comment by curran** [comment]
Infinite scroll does seem to work actually. Amazing!

**Comment by curran** [comment]
@stushurik One thing I notice here: when I fork a viz, the set of viz previews shown on the home page does not update.

I think this is desired behavior, in a way. It allows the server to mediate the set based on querying against access control rules.

I wonder, though, if we could achieve that feature of real-time query result set updates. I'll track that separately.

**Review by curran — APPROVED** [review]
Looks great!

**Pull request merged**